### PR TITLE
add code for pinging Clearcut

### PR DESCRIPTION
**Pull request by elharo**
@suztomo toward #1619
@chanseokoh I would also appreciate your review on this since a lot of it is adapted from CT4E code.

### CODE REVIEW

File: linkage-monitor/src/test/java/com/google/cloud/tools/dependencies/linkagemonitor/AnalyticsTest.java [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+  public void testQueuePingArguments_validEventName() {
+    analytics.queuePing("eventName");
+  }
+
+  @Test
+  public void testQueuePingArguments_nullEventName() {
+    try {
+      analytics.queuePing(null);
+      fail();
+   } catch (NullPointerException ex) {
```
Comment (suztomo) [inline linkage-monitor/src/test/java/com/google/cloud/tools/dependencies/linkagemonitor/AnalyticsTest.java:93]: Fix indent

File: linkage-monitor/src/test/java/com/google/cloud/tools/dependencies/linkagemonitor/AnalyticsTest.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+      assertEquals("eventName null", ex.getMessage());
+    }
+  }
+
+  @Test
+  public void testQueuePingArguments_emptyEventName() {
+    try {
+      analytics.queuePing("");
+      fail();
+   } catch (IllegalArgumentException ex) {
```
Comment (suztomo) [inline linkage-monitor/src/test/java/com/google/cloud/tools/dependencies/linkagemonitor/AnalyticsTest.java:103]: Fix indent.

File: linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+    PingEvent(String eventName, ImmutableMap<String, String> metadata) {
+      Preconditions.checkArgument(!Strings.isNullOrEmpty(eventName), "eventName null or empty");
+      Preconditions.checkNotNull(metadata, "metadata is null");
+      this.eventName = eventName;
+      this.metadata = metadata;
+    }
+  }
+
+  @VisibleForTesting
+  String jsonEncode(PingEvent event) {
```
Comment (suztomo) [inline linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java:168]: Would you add a test for this method?

File: linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+    Gson gson = new Gson();
+
+    Map<String, String> desktopClientInfo = new HashMap<>();
+    desktopClientInfo.put("os", System.getProperty("os.name"));
```
Comment (suztomo) [inline linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java:172]: Linkage Monitor does not target for desktops.

File: linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+    String sourceExtensionJsonString = gson.toJson(sourceExtension);
+    
+    Map<String, Object> logEvent = new HashMap<>();
+    logEvent.put("event_time_ms", System.currentTimeMillis());
+    logEvent.put("sequence_position", sequencePosition++);  
+    logEvent.put("source_extension_json", sourceExtensionJsonString);
+    
+    Map<String, Object> root = new HashMap<>();
+    root.put("log_source", "CONCORD");
+    root.put("request_time_ms", System.currentTimeMillis());
```
Comment (suztomo) [inline linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java:204]: `System.currentTimeMillis()` is banned in go/goodtime/banned#java.lang.System.currentTimeMillis()

> Please use TimeSource.now() instead (which returns a Instant)

File: linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java (hunk 4 above)
Comment (elharo) [reply linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java:176]: I think this is required for logging though. @chanseokoh Do you recall?

File: linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java (hunk 5 above)
Comment (elharo) [reply linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java:208]: No, that class is only available in google3, not open source, at least not yet.

File: linkage-monitor/src/test/java/com/google/cloud/tools/dependencies/linkagemonitor/AnalyticsTest.java (hunk 1 above)
Comment (elharo) [reply linkage-monitor/src/test/java/com/google/cloud/tools/dependencies/linkagemonitor/AnalyticsTest.java]: done

File: linkage-monitor/src/test/java/com/google/cloud/tools/dependencies/linkagemonitor/AnalyticsTest.java (hunk 2 above)
Comment (elharo) [reply linkage-monitor/src/test/java/com/google/cloud/tools/dependencies/linkagemonitor/AnalyticsTest.java]: done

File: linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java (hunk 3 above)
Comment (elharo) [reply linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java:172]: done

File: linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java (hunk 5 above)
Comment (suztomo) [reply linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java:212]: Oh, ok. Thanks.

File: linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java (hunk 4 above)
Comment (suztomo) [reply linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java:180]: I meant the variable name.

```suggestion
    Map<String, String> osInfo = new HashMap<>();
    osInfo.put("os", System.getProperty("os.name"));
```

I don't have any concern on sending operating system names but I expect they are all Linux.

File: linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+
+  @VisibleForTesting
+  String jsonEncode(PingEvent event) {
+    Gson gson = new Gson();
+
+    Map<String, String> desktopClientInfo = new HashMap<>();
+    desktopClientInfo.put("os", System.getProperty("os.name"));
+    
+    Map<String, Object> clientInfo = new HashMap<>();
+    clientInfo.put("client_type", "DESKTOP");
```
Comment (suztomo) [inline linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java:183]: What's proper value for things like Github Actions?

Comment (suztomo) [reply linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java:183]: Found Chanseok's message https://groups.google.com/a/google.com/g/concord-dev/c/-wb7MzRpSv4/m/hZ-tD0WrAgAJ with clientanalytics.proto. I don't find a suitable value among the file.

File: linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java (hunk 4 above)
Comment (elharo) [reply linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java:180]: on line 184 note that this value is stored in a field with the name "desktop_client_info" as defined by the proto

File: linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java (hunk 6 above)
Comment (elharo) [reply linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java:183]: I don't see one better than DESKTOP. I don't think they call out webhooks like github actions separately.

**Review by suztomo — APPROVED** [review]

File: linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+import java.util.Map;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Report events to Google Analytics.
```
Comment (chanseokoh) [inline linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java:40]: Definitely you are not using [Google Analytics](https://en.wikipedia.org/wiki/Google_Analytics). We've migrated away from it in CT4E, and that's gone. "Analytics" as a general term may work, but just don't call it "Google Analytics."

File: linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+    Preconditions.checkNotNull(metadataKey, "metadataKey null");
+    Preconditions.checkArgument(!metadataKey.isEmpty(), "metadataKey empty");
+    Preconditions.checkNotNull(metadataValue, "metadataValue null");
+    Preconditions.checkArgument(!metadataValue.isEmpty(), "metadataValue empty");
+
+    queuePing(eventName, ImmutableMap.of(metadataKey, metadataValue));
+  }
+
+  /**
+   * Sends a usage metric to Google Analytics.
```
Comment (chanseokoh) [inline linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java:126]: ditto

File: linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+    root.put("request_time_ms", System.currentTimeMillis());
+    root.put("client_info", clientInfo);
+    root.put("log_event", Collections.singletonList(logEvent));
+
+    return gson.toJson(root);
+  }
+
+  @VisibleForTesting
+  static String getAnonymizedClientId() {
+    return "clientId"; // TODO fill in with what????";
```
Comment (chanseokoh) [inline linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java:221]: I don't recall a specific and exact requirement for "client_install_id", but basically, in CT4E, this is for distinguishing and counting different "users" or "clients." In CT4E, we just generated a UUID that will be persisted in an Eclipse instance, hopefully matching one unique "user" per Eclipse installation (or instance?).

File: linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+      } 
+    }
+  }
+
+  @VisibleForTesting
+  static class PingEvent {
+    @VisibleForTesting final String eventName;
+    @VisibleForTesting final ImmutableMap<String, String> metadata;
+
+    PingEvent(String eventName, ImmutableMap<String, String> metadata) {
```
Comment (chanseokoh) [inline linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java:167]: `private` or `@VisibleForTesting`?

File: linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+    this.pingEventQueue = concurrentLinkedQueue;
+    this.sendAnalytics = sendAnalytics;
+  }
+
+  // TODO any synchronization issues here? triple check that this makes sense.
+  private void start() {
+    Runnable emptyQueue = new Runnable() {
+      @Override
+      public void run() {
+        // drain the queue; GA can handle our traffic without buffering
```
Comment (chanseokoh) [inline linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java:81]: "GA" --> backend?

File: linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java [hunk 12]
```diff
@@ -0,0 +1,223 @@
...
+        PingEvent event = pingEventQueue.poll();
+        while (event != null) {
+          if (sendAnalytics && event != null) {
```
Comment (chanseokoh) [inline linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java:84]: This condition is also in `ping()` and redundant.

File: linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java [hunk 13]
```diff
@@ -0,0 +1,223 @@
...
+  // TODO is this a good value?
+  private static final int DELAY = 773; // a prime number
+
+  private static final Logger logger = Logger.getLogger(Analytics.class.getName());
+
+  private static final String FIRELOG_COLLECTION_URL =
+      "https://firebaselogging-pa.googleapis.com/v1/firelog/legacy/log";
+
+  // Don't change the value; this name is used as an originating "application" of usage metrics.
+  private static final String METRICS_NAME = "cloud-opensource-java-linkage-monitor";
```
Comment (chanseokoh) [inline linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java:53]: This is not used anywhere.

File: linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java [hunk 14]
```diff
@@ -0,0 +1,223 @@
...
+  
+  // for Concord Clearcut
```
Comment (chanseokoh) [inline linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java:55]: "Clearcut" --> "Firelog". We also migrated away from Clearcut to Firelog.

File: linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+    ScheduledExecutorService scheduler = MoreExecutors.getExitingScheduledExecutorService(
+        executor, 10, TimeUnit.SECONDS);
+
+    scheduler.scheduleAtFixedRate(emptyQueue, DELAY, DELAY, TimeUnit.MILLISECONDS);
+  }
+
+  static synchronized Analytics getInstance(boolean sendAnalytics) {
+    if (instance == null) {
+      String collectionUrl = null;
+      if (!PlaceholderConstants.FIRELOG_API_KEY.startsWith("@")) {
```
Comment (suztomo) [inline linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java:105]: I don't see `FIRELOG_API_KEY` any more. I remember something was to replace compile-time value. Where did it go?

Comment (elharo) [reply linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java:104]: Already checked in https://github.com/GoogleCloudPlatform/cloud-opensource-java/blob/master/linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/PlaceholderConstants.java

File: linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java (hunk 12 above)
Comment (elharo) [reply linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java]: removed null check

File: linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java (hunk 11 above)
Comment (elharo) [reply linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java]: Clearcut?

File: linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java (hunk 14 above)
Comment (elharo) [reply linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java:55]: I thought we were using the Firelog ingestion service to send data into Clearcut. Are we instead using Firelog without Clearcut?

File: linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java (hunk 10 above)
Comment (elharo) [reply linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java:168]: done

File: linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java (hunk 15 above)
Comment (suztomo) [reply linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java:104]: Thank you.

File: linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java (hunk 14 above)
Comment (chanseokoh) [reply linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java:55]: Ah, you may be right. Honestly, I don't really know how the backend works.

**Comment by suztomo** [comment]
We didn't go for Clearcut.

**Pull request closed without merging**